### PR TITLE
New padding options for `Conv` and `ConvTranspose`

### PR DESCRIPTION
**Pull request by ChenAo-Phys**
According to #638, some changes are made to the padding in `Conv` and `ConvTranspose`.

The `padding` argument now supports three string inputs, "VALID", "SAME", or "SAME_LOWER", similar to [`jax.lax.conv_general_dilated`](https://jax.readthedocs.io/en/latest/_autosummary/jax.lax.conv_general_dilated.html).

Furthermore, there is a new argument `padding_mode` which accepts string inputs including "ZEROS" (default), "REFLECT", "REPLICATE", and "CIRCULAR", similar to [PyTorch](https://pytorch.org/docs/stable/generated/torch.nn.Conv2d.html).

To make the "SAME" or "SAME_LOWER" padding work for `ConvTranspose`, the definition of `output_padding` is slightly changed to make `output_size = (input_size + output_padding) / stride` in these cases.

Only "ZEROS" and "CIRCULAR" are implemented in `ConvTranspose`. "CIRCULAR" further requires `output_padding == 0` 
 and `padding == "SAME"` or `"SAME_LOWER"` to simplify it, which is the typical case when people use "CIRCULAR" padding.

### CODE REVIEW

File: equinox/nn/_conv.py [hunk 1]
```diff
@@ -80,6 +116,8 @@ def __init__(
             concatenating the results along the output channel dimension.
             `in_channels` must be divisible by `groups`.
         - `use_bias`: Whether to add on a bias after the convolution.
+        - `padding_mode`: One of the following strings specifying the padding values.
+            `'ZEROS'` (default), `'REFLECT'`, `'REPLICATE'`, or `'CIRCULAR'`.
```
Comment (patrick-kidger) [inline equinox/nn/_conv.py:120]: Can you explain what each of these options does?

File: equinox/nn/_conv.py [hunk 2]
```diff
@@ -134,21 +174,34 @@ def __init__(
...
         self.groups = groups
         self.use_bias = use_bias
+        self.padding_mode = _padding_mode_init(padding_mode)
+
+    def _nonzero_pad(self, x: Array) -> Array:
+        if isinstance(self.padding, str):
+            rhs_shape = tuple(
+                d * (k - 1) + 1 for k, d in zip(self.kernel_size, self.dilation)
+            )
+            padding = lax.padtype_to_pads(
```
Comment (patrick-kidger) [inline equinox/nn/_conv.py:188]: I don't think this is public API, unfortunately. ([It's not listed here, anyway.](https://jax.readthedocs.io/en/latest/jax.lax.html)) Maybe we can ask to add it to the public API for JAX?

File: equinox/nn/_conv.py [hunk 3]
```diff
@@ -134,21 +174,34 @@ def __init__(
...
+                x.shape[2:], rhs_shape, self.stride, self.padding
```
Comment (patrick-kidger) [inline equinox/nn/_conv.py:189]: Why `[2:]`?
`x` has shape `(in_channels, dim_1, ..., dim_N)`, so this is trimming off `dim_1` as well.

File: equinox/nn/_conv.py [hunk 4]
```diff
@@ -393,27 +466,72 @@ def __init__(
...
             self.bias = None
 
+        padding = _padding_init(padding, num_spatial_dims)
+        padding_mode = _padding_mode_init(padding_mode)
+        if padding_mode in ("REFLECT", "REPLICATE"):
+            raise NotImplementedError(
+                "'REFLECT' or 'REPLICATE' padding mode is not implemented"
+            )
+        elif padding_mode == "CIRCULAR":
+            if np.any(output_padding):
```
Comment (patrick-kidger) [inline equinox/nn/_conv.py:476]: Nit: I prefer avoiding implicit-cast-to-bool. (Here for the int->bool conversion on each element.) Can this be `if any(x !=0 for x in output_padding)`?

**Review by patrick-kidger — COMMENTED** [review]
Thank you for the contribution! This is excellent; thank you for implementing so much.

My main two comments are:
1. Can you add some more documentation to the docstrings? (E.g. what does it mean to use `padding_mode == "CIRCULAR"` in `ConvTranspose`?)
2. Can you add some tests? I think I spot a few cases where the wrong thing is going to happen at the moment (I've commented on a couple of them below.)

File: equinox/nn/_conv.py (hunk 3 above)
Comment (ChenAo-Phys) [reply equinox/nn/_conv.py:189]: A batch dimension is expanded in line 226, so the shape is `(batch, in_channels, dim_1 , ...)`.

Comment (patrick-kidger) [reply equinox/nn/_conv.py:189]: Ah, makes sense! Thanks.

File: equinox/nn/_conv.py (hunk 2 above)
Comment (ChenAo-Phys) [reply equinox/nn/_conv.py:191]: Could you raise a new issue in JAX? I think your proposal is much more likely to be accepted

**Comment by ChenAo-Phys** [comment]
> Thank you for the contribution! This is excellent; thank you for implementing so much.
> 
> My main two comments are:
> 
> 1. Can you add some more documentation to the docstrings? (E.g. what does it mean to use `padding_mode == "CIRCULAR"` in `ConvTranspose`?)
> 2. Can you add some tests? I think I spot a few cases where the wrong thing is going to happen at the moment (I've commented on a couple of them below.)

Some more documentation and tests have been added according to your suggestion. Thanks a lot!

File: equinox/nn/_conv.py (hunk 2 above)
Comment (patrick-kidger) [reply equinox/nn/_conv.py:191]: Done! We'll see what they say.

**Comment by patrick-kidger** [comment]
Looks like the pre-commit hooks are failing. (See CONTRIBUTING.md for setting these up locally.)

Modulo that detail, this PR LGTM! Thank you for putting it together, it'll be a great addition to have. Let me know once you've got the tests passing and I'll merge this PR.

**Comment by ChenAo-Phys** [comment]
Some type hints have been fixed.

There seems to be something wrong with the test. It always reports 
`FAILED tests/test_errors.py::test_tracetime - Failed: DID NOT WARN. No warnings of type (<class 'UserWarning'>,) were emitted.`

**Comment by patrick-kidger** [comment]
Rebase on top of the latest `dev`, I fixed this one a few days ago :)

**Comment by patrick-kidger** [comment]
And merged! Thank you for the contribution. this is great to have. This will appear in the next release of Equinox :)

**Comment by ChenAo-Phys** [comment]
> And merged! Thank you for the contribution. this is great to have. This will appear in the next release of Equinox :)

Nice! Also thanks for your time!
